### ISSUE TITLE
[hdf5] header file (alps/hdf5/archive.hpp) conflicts with ALPS

### ISSUE THREAD

**Issue by wistaria** [ALPSCore/ALPSCore#582]
Name conflict (alps/hdf5/archive.hpp) between ALPS and ALPSCore prevents from installing the both libraries into the same directly, e.g. under system-wide /usr/include.

**Comment by egull** [comment]
Yes. Not sure what to do about that one or if we're even going to try to address it. Ideally people will switch over to the core library, but we don't have the manpower to take care of the legacy codes so ALPS will be around for some time. I think it's worse than described here: if there are ALPS headers lurking in /usr/include and ALPSCore is installed somewhere else, the compilers may mix and match headers with unpredictable outcome.

**Comment by galexv** [comment]
An option would be to rename all header from `<alps/abc.hpp>` to `<alps/core/abc.hpp>` (this has to be a new major version release, because it breaks the interface) and offer a migration tool: a script that finds all alpscore headers by `^( *# *include *<alps/)` regex and changes them to `${1}core/` .

**Comment by egull** [comment]
Yeah that's not very practical. Neither is renaming the old headers to .../alps/legacy/... I don't see a good way to fix this, other than to tell people not to install legacy alps and the new version system wide.

**Comment by galexv** [comment]
Then it should be documented: i am placing this issue in the Documentation epic.

**Comment by mwallerb** [comment]
Does legacy ALPS set some sort of preprocessor macro that ALPS core does not use? In this case, we could at least check for the presence of that macro and complain ...?

**Comment by egull** [comment]
Dicey. A lot in new alps is similar to what it was in old alps, but the header names are mostly different. config.hpp was config.h, for example. We could possibly check for the alps/config.h header in the standard path during cmake and bail if it exists? That one is an indication that there's an old ALPS lying around. Something like this:


  CHECK_INCLUDE_FILE_CXX("alps/config.h" HAVE_OLD_ALPS)
  IF(HAVE_OLD_ALPS)
    message( FATAL_ERROR "Old alps exists in your system. Please do not compile alpscore with old alps in your include path." )
  ENDIF()

**Comment by mwallerb** [comment]
@egull, well we just need a single macro in, say, `alps/config.h` that is unused in ALPS core.

Then we can just check somewhere deep in alps core
```C
    #ifdef ALPS_SOME_LEGACY_MACRO
        #error "Croaky-croak: You attempt to run ALPS and ALPS core side-by-side"
    #endif
```

I would do this in a header rather than CMake, b/c this way we catch it at the config actually used by the user.

**Comment by egull** [comment]
For that to work you'd have to find a macro that every ALPS header defines but that's not in ALPSCore, and then every alpscore header would have to test for it in a part that is guaranteed not to be overwritten by alps. I don't know how to do that. For instance, the macro ALPS_NGS_USE_NEW_ALEA is a giveaway that someone used the old alea config.h. But then where would you check for it? not in config.hpp, because that one most likely got overridden and replaced by config.hpp. Even if it didn't, you have no guarantee that config.hpp is included after config.h. So... I have no idea how to do this right. Do you know?
